plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -60,120 +64,110 @@ becomes
+-- the application, if possible.
+fullyApplyAndBetaReduce ::
+  forall tyname name uni fun ann.
+  (InliningConstraints tyname name uni fun) =>
+  -- | The variable
+  VarInfo tyname name uni fun ann ->
+  -- | The arguments
+  AppContext tyname name uni fun ann ->
+  InlineM tyname name uni fun ann (Maybe (Term tyname name uni fun ann))
+fullyApplyAndBetaReduce info = go (varBody info) (varArity info)
+  where
+    go ::
+      Term tyname name uni fun ann ->
+      Arity tyname name ->
+      AppContext tyname name uni fun ann ->
+      InlineM tyname name uni fun ann (Maybe (Term tyname name uni fun ann))
+    go acc arity args = case (arity, args) of
+      -- success
+      ([], _) -> pure . Just $ fillAppContext acc args
+      (TermParam param : arity', TermAppContext arg _ args') ->
+        ifM
+          (safeToSubst param arg)
+          ( go
+              (termSubstNames (\n -> if n == param then Just arg else Nothing) acc)
+              arity'
+              args'
+          )
+          (pure Nothing)
+      (TypeParam param : arity', TypeAppContext arg _ args') ->
+        go
+          (termSubstTyNames (\n -> if n == param then Just arg else Nothing) acc)
+          arity'
+          args'
+      _ -> pure Nothing
+
+    -- Checks whether it is safe (i.e., preserves semantics) to perform beta reduction,
+    -- namely, turning `(\a -> body) arg` into `body [a := arg]`.
+    safeToSubst :: name -> Term tyname name uni fun ann -> InlineM tyname name uni fun ann Bool
+    safeToSubst n = effectSafe (varBody info) Strict n <=< checkPurity
+
+-- | Consider whether to inline an application.
+inlineSaturatedApp ::
+  forall tyname name uni fun ann.
+  (InliningConstraints tyname name uni fun) =>
+  Term tyname name uni fun ann ->
+  InlineM tyname name uni fun ann (Term tyname name uni fun ann)
+inlineSaturatedApp t
+  | (Var _ann name, args) <- splitApplication t =
+      gets (lookupVarInfo name) >>= \case
+        Just varInfo -> fullyApplyAndBetaReduce varInfo args >>= \case
+          Just fullyApplied -> do
+            let def = varDef varInfo
+                -- Inline only if the size is no bigger than not inlining.
+                sizeIsOk = termSize fullyApplied <= termSize t
                 -- The definition itself will be inlined, so we need to check that the cost
                 -- of that is acceptable. Note that we do _not_ check the cost of the _body_.
                 -- We would have paid that regardless.
-                -- Consider e.g. `let y = \x. f x`. We pay the cost of the `f x` at every call
-                -- site regardless. The work that is being duplicated is the work for the lambda.
-                defCostOk = costIsAcceptable def
+                -- Consider e.g. `let y = \x. f x`. We pay the cost of the `f x` at
+                -- every call site regardless. The work that is being duplicated is
+                -- the work for the lambda.
+                costIsOk = costIsAcceptable def
```

Comment:
Hmmm, maybe we can actually glue together multiple pieces of argumentation. Suppose that `costIsAcceptable def` is true. Then we are going to:
- Inline the variable. This is a cost saving since `costIsAcceptable def` is true
- Do beta reduction. This is a cost saving because we are following our usual inlining heuristics, which should ensure that the transformation is a cost saving

So I think as long as we do match the existing inlining heuristics for step 2, the logic does actually check out? It's a bit subtle, though, worth writing down carefully.

Reply:
SGTM